Import hooks

nickmoreton opened the pull request:
# Import Hooks

This work allows a developer to configure one or more XML tags within the XML <item> tag to be used as a key/indicator to cache the complete <item> for use during the import. 

At a later stage of the import the cached items can be queried and the data for that item is used to create foreign key objects on the Wagtail page before the import process ends.

Ticket URL: https://projects.torchbox.com/projects/wordpress-to-wagtail-importer-package/tickets/73

---

- Testing
    - [x] CI passes
    - [x] If necessary, tests are added for new or fixed behaviour
    - [x] These changes do not reduce test coverage
- Documentation.
    - [x] This PR adds or updates documentation

nickmoreton committed:
add wp_post_meta field

nickmoreton committed:
make migrations

nickmoreton committed:
add xml boilerplate

nickmoreton committed:
add import hooks module

nickmoreton committed:
add method to cache item tags

save

nickmoreton committed:
add method to save post meta to page object

save

nickmoreton committed:
extend tests for post meta is saved

nickmoreton committed:
extend import hooks config

nickmoreton committed:
add mock patch for tetsing the items are cached

add tests for items_cache methods

tests import hooks

tests docs

nickmoreton committed:
add methods for processing cached items per page

code docs

save

hooks docs

save

nickmoreton committed:
add call to items_cache.process

nickmoreton committed:
add in call to ItemsCache.process

nickmoreton committed:
Revert "make migrations"

This reverts commit 983e311e2f34b98cd2bd7c513741422e2adef6a1.

nickmoreton committed:
redo make migrations

nickmoreton committed:
add import hooks documentation

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        }
+    }
+    """
+    return getattr(settings, "WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE", {})
```
**Issue**: This is using a function to return something unchanged from settings. It's not dynamic. Why don't we just repeat `getattr(settings, "WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE", {})` wherever it's needed? This seems to be what the Wagtail project does.

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        and stored in the attachment attribute.
+        """
+
+        for hook in import_hooks_xml_items_to_cache():
```
**Issue**: This is looping over a dict, which will return the keys. Can we explicitly loop over the keys instead, for clarity?

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+            "wp:post_type": "attachment",
+        }
+        and stored in the attachment attribute.
+        """
```
**Question**: Does this `__init__` method actually do what this docstring says? Is it just describing what will be done elsewhere, e.g. by `WordpressImporter.cache_item_tags`? If so maybe this just belongs in the docs files.

nickmoreton replied:
I removed the incorrect doc string content

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        """
+
+        for hook in import_hooks_xml_items_to_cache():
+            setattr(self, hook, [])
```
**Suggestion**: instead of using an object here, could we use [a defaultdict?](https://docs.python.org/3/library/collections.html#collections.defaultdict). Example usage:

```python
>>> from collections import defaultdict
>>> cache = defaultdict(list)
>>> cache
defaultdict(<class 'list'>, {})
>>> dict(cache)
{}
>>> cache["foo"].append(1)
>>> dict(cache)
{'foo': [1]}
```

nickmoreton replied:
We decided to leave this as is for the moment.

nimasmi commented on wagtail_wordpress_import/importers/wordpress.py:
```diff
@@ -65,6 +71,13 @@ def run(self, *args, **kwargs):
+                # which is a XML item that is a media type
+                for hook in import_hooks_xml_items_to_cache():
+                    if item.get("wp:post_type") == hook:
+                        self.cache_item_tags(item, hook)
```
**Suggestion** We can avoid a loop here with:

```suggestion
                post_type = item.get("wp:post_type")
                if post_type in import_hooks_xml_items_to_cache():
                   self.cache_item_tags(item, post_type)
```

However, I don't think this `if` statement should stay here. I have also suggested a change to `cache_item_tags`, to move it to the cache. So if that is accepted, here we would have to send the data straight to a method on the cache:

```suggestion
                post_type = item.get("wp:post_type")
                self.item_cache.cache_item(item, post_type)
```

(and reword the comments above)

nimasmi commented on wagtail_wordpress_import/importers/wordpress.py:
```diff
@@ -257,6 +273,14 @@ def connect_page_categories(self, page, category_model, item):
+            hooks_attr.append(item)
+
+        return self.items_cache
+
```
**Suggestion**: Ditch this, and move this method to the cache.

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        """
+
+        for hook in import_hooks_xml_items_to_cache():
+            setattr(self, hook, [])
```
**Suggestion**: Because elsewhere we check for membership before appending, a set will do that better.

```suggestion
            setattr(self, hook, set())
```

nickmoreton replied:
@nimasmi I am having an issue here when creating the hook as a set() and trying to add the item which is a dict.

```python
def add_item_to_cache(self, hook, item):
        """Add an item dict to the cached hook if not already added"""
        hook = getattr(self, hook)
        hook.add(item)
```
I tried `hook.add(json.dumps(item))` which works but then the other logic to retrieve the items will need to be changed. Just wondering if thats expected.

nimasmi replied:
Yeah, sorry. This was a result of my hastily submitting the code review at the end of Wednesday. If it's stored as JSON, it will need unpacking elsewhere.

nimasmi commented on wagtail_wordpress_import/xml_boilerplate.py:
```diff
@@ -0,0 +1,40 @@
+from io import StringIO
```
**Question**: Is this file purely for tests? If so, move it to either a tests folder of the importers app, or to the tests app.

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        for hook, actions in import_hooks_xml_items_to_cache().items():
+            if hook in items_cache:
+                return actions["FUNCTION"], actions["DATA_TAG"]
+        return None, None
```
**Suggestion**: You are performing a check that this hook/post_type is in the list of things to cache. Whether to do that check in `WordpressImporter.run()`, or here, I wasn't at first certain, but I think now that by keeping a track only here of the keys to be cached, it's cleaner. I may have made some confusing suggestions though because refactoring is hard in the GitHub PR UI. Still, we don't need to check whether the post_type is a cacheable hook more than once. I suggest doing it all here.

```suggestion
        return None, None

    def cache_item(self, item, hook):
        if hasattr(self, hook):
            hooks_attr = getattr(self, hook)
            if item not in hooks_attr:
                hooks_attr.append(item)
```

Or better, if you change the lists to sets:

```suggestion
        return None, None

    def cache_item(self, item, hook):
        if hasattr(self, hook):
            hooks_attr = getattr(self, hook)
            hooks_attr.add(json.dumps(item))
```

This uses the JSON representation of the dict, because dicts are mutable, and so not hashable. It does permit us later to refactor to use a cache backend though, rather than keeping this ItemCache object in memory.

NB there's no return value here, as it wasn't used where called originally in https://github.com/torchbox/wagtail-wordpress-import/pull/84/files#diff-3ab0aa3f3132ef5666a66ab62fad501c8ee19c813c94c72720522117ed6a98acR79

nickmoreton replied:
@nimasmi Just read this after posting the earlier question. I'll see whats involved to do it this way.

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        """Return all the cached attributes added from the current config"""
+        return self.__dict__
+
+    def process(self, imported_pages, items_cache):
```
**Issue**: It looks like `items_cache` and `self` are the same thing. Do we need this extra argument?

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+                import_string(func)(page, data, items_cache)
+
+    @staticmethod
+    def _get_hook_handler_data(items_cache):
```
**Suggestion**: this is a staticmethod that takes a class instance as an argument. It should be a plain method.

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,86 @@
+        """Get the hook function and data to process"""
+        for hook, actions in import_hooks_xml_items_to_cache().items():
+            if hook in items_cache:
+                return actions["FUNCTION"], actions["DATA_TAG"]
```
**Issue**: This returns after the first positive match, but the rest of this class suggests that it's going to need to return a list of `(func, data)` tuples.

See also the for loop at https://github.com/torchbox/wagtail-wordpress-import/pull/84/files#diff-60d6d8cf68ae5d46f2f79910bde207af85ed5f28013bc2e8874a85461649194dR77 that doesn't use `hook` inside the loop.

nimasmi replied:
See #96 for a solution to this.

nimasmi reviewed (changes requested):
Hi Nick. There was so much that I wanted to discuss about this that I think we should meet to plan it tomorrow. I began with suggestions, but really GitHub's PR UI isn't suitable for this discussion.

nickmoreton committed:
remove import_hooks_xml_items_to_cache() and refactor in the change

nickmoreton committed:
add method add_item_to_cache()

nickmoreton committed:
remove cache_item_tags method and implement add_item_to_cache mthod on ItemsCache

nickmoreton committed:
move xml_boilerplate.py and refactor tests

nickmoreton committed:
use self.get_cached_items() in process method

items cache and self are the same thing

nickmoreton committed:
refactor import_hooks.py

**Issue**: It looks like `items_cache` and `self` are the same thing. Do we need this extra argument?

refactor process method

nickmoreton committed:
only run hooks if WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE is set

nimasmi commented on docs/import_hooks.md:
```diff
@@ -0,0 +1,164 @@
+
+A full Wordpress XML export of a complete Wordpress website will contain all the entities within the database but as it's a text file the relationships between the entities are not ideally represented for the import process.
+
+**For example.** A blog post might include a header image thats not part of the body content in the Wordpress XML `<item>` tag for the post. It would be referenced by an ID in the `<wp:postmeta>` tag.
```
**Suggestion**: Wording

```suggestion
**For example**, a blog post might include a header image that's not part of the body content in the Wordpress XML `<item>` tag for the post. The image would be elsewhere in the XML, referenced by an ID in the post item's `<wp:postmeta>` tag.
```

nimasmi commented on docs/import_hooks.md:
```diff
@@ -0,0 +1,164 @@
+
+## What are import hooks useful for?
+
+A full Wordpress XML export of a complete Wordpress website will contain all the entities within the database but as it's a text file the relationships between the entities are not ideally represented for the import process.
```
**Suggestion** (nitpick): Casing and punctuation

```suggestion
A full XML export of a complete WordPress website will contain all the entities within the database, but as it's a text file the relationships between the entities are not ideally represented for the import process.
```

nimasmi commented on docs/import_hooks.md:
````diff
@@ -0,0 +1,164 @@
+</item>
+```
+
+This `<item>` XML tag above is not included when an import is run because we only import `<item>` tags that represent a blog post or page to be created in Wagtail. These are items with a `<wp:post_type>` of post or page by default. The sample XML above represents an `attachment` type.
````
**Suggestion**: Wording

```suggestion
The type of an XML `<item>` element is given by its `<wp:post_type>` tag. The example item above represents an "attachment" type. This is not included by default when an import is run, because we only import items of the "post" or "page" type, representing a blog post or page to be created in Wagtail.
```

nimasmi commented on docs/import_hooks.md:
```diff
@@ -0,0 +1,164 @@
+
+## Adding an import hook
+
+The package doesn't include any import hooks by default. You can add your own import hooks by adding a dictionary the the  `WAGTAIL_WORDPRESS_IMPORT_HOOKS` setting and defining a function thats will be called and passed the Page model and data to be imported.
```
**Question**: Is this an agreed position? Do we know enough about WordPress defaults to know whether thumbnails, header images, and authors are standard elements?

nimasmi replied:
**Suggestion**: Typos

```suggestion
The package doesn't include any import hooks by default. You can add your own import hooks by adding a dictionary as the  `WAGTAIL_WORDPRESS_IMPORT_HOOKS` setting and defining an XML post type to identify it, and a function to process the data. The function takes a page instance, and a cache of the tag data to be processed, as arguments.
```

nimasmi commented on docs/import_hooks.md:
````diff
@@ -0,0 +1,164 @@
+```python
+{
+    ...
+    # each dict item is a nested dict that represents the XML item tags tree 
````
**Suggestion**: Correct description of data structure

```suggestion
    # each dict item is a nested list of dicts that represents the XML item tags tree 
```

nimasmi commented on docs/import_hooks.md:
````diff
@@ -0,0 +1,164 @@
+}
+```
+
+The `FUNCTION` key is the dotted path to the function that will be called. The function should accept the Page model and the data to be imported as arguments. The function also needs to be passed an argument of `items_cache` whcih is the cache of all the items that have been imported with the `<wp:post_type>attachment</wp:post_type>`
````
**Suggestion**: typos

```suggestion
The `FUNCTION` key is the dotted path to the function that will be called. The function should accept the Page model and the data to be imported as arguments. The function also needs to be passed an argument of `items_cache` which is the cache of all the items that have been imported with the tag `<wp:post_type>attachment</wp:post_type>`
```

nimasmi commented on docs/import_hooks.md:
```diff
@@ -0,0 +1,164 @@
+    attachments = items_cache.get("attachment")
+
+    # the code below would be your own code to process the header image,
+    # you can perform any type of python/Wagtail logic here and make use 
```
**Suggestion**: Capitalisation

```suggestion
    # you can perform any type of Python/Wagtail logic here and make use 
```

nimasmi commented on docs/import_hooks.md:
```diff
@@ -0,0 +1,164 @@
+    # of thumbnail_id and attachments variables.
+    image_url = None
+
+    for attachment in attachments:
+        if (
+            attachment.get("wp:post_id")
+            and attachment.get("wp:post_id") == thumbnail_id
+        ):
+            image_url = attachment.get("guid")
+
+    if image_url:
+        image = get_or_save_image(image_url)
+        page.header_image = image
+        page.save()
```
**Suggestion**: I know this is only example code, but I'm interested that we iterate over the pages (I presume), when calling this function on each imported page, but then within this function iterate over every single attachment. If we expect the cached attachments to contain things we can query on the page (as we're using JSONField), then would the following (pseudocode) work?

```python
for attachment in attachments:
    thumbnail_id = (some way to get thumbnail_id)
    lookup = f"wp_post_meta__{data}"  # i.e. construct a Django filter kwarg
    pages = ImportedPage.objects.filter(**{lookup: thumbnail_id})
    if pages.exists():
        image = # do something to fetch and store the image
        pages.update(header_image=image)
```

i.e. one iteration overall, and two queries per attachment, no nested forloop

nickmoreton replied:
I will try this locally and thanks, thats made the JSONField query make sense to me now :)

nimasmi replied:
Thanks. On second thought, I'd change my line to assume a `pages` queryset was provided, rather than presupposing the use of ImportedPage here:

```python
- pages = ImportedPage.objects.filter(**{lookup: thumbnail_id})
+ pages = pages.filter(**{lookup: thumbnail_id})
```

nickmoreton replied:
In import authors the user example is now but it's not working here because the wp_post_meta is stored in a different way with a new function.

```python
def header_image(imported_pages, data, items_cache):
    """use this for the import authors version"""
    lookup = f"wp_post_meta__{data}"
    for attachment in items_cache:
        thumbnail_id = attachment.get("wp:post_id")
        pages = imported_pages.filter(**{lookup: thumbnail_id})
        if pages.exists():
            image = get_or_save_image(attachment.get("guid"))
            pages.update(header_image=image)
            print("Attaching header images to pages:", pages)
```

nimasmi commented on docs/import_hooks.md:
````diff
@@ -0,0 +1,164 @@
+        ImageChooserPanel("header_image"),
+        ...
+    ]
+```
````
**Praise**: These are good docs, @nickmoreton, thanks.

nickmoreton replied:
Cheers! With the help of GitHub Copilot, it's pretty amazing!

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,42 @@
+    def process(self, imported_pages):
+        """Run all hooks in the config for each page
+        The function defined in the config key FUNCTION will be imported
+        and run on each page"""
```
**Comment**: This is probably the point I'm referring to with my discussion in the docs about the example header import function. Could we leave the iteration logic until this function, and let it query the pages, rather than calling it once per page from elsewhere. It might make sense for some data to be called once per page, and for others to iterate over the cached data instead, depending on the many-to-many/one-to-many or left/right join nature of the relationships.

It's probably outside the scope of this PR.

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,42 @@
+        and run on each page"""
+        for page in imported_pages:
+            func, data = self._get_hook_handler_data()
+            import_string(func)(page, data, self.__dict__)
```
**Issue**: This code, and the `self._get_hook_handler_data` method only cope with a single entry in `WORDPRESS_IMPORT_HOOKS_ITEMS_TO_CACHE`. See the `return` statement within a forloop on line 41 below.

I originally reported this at https://github.com/torchbox/wagtail-wordpress-import/pull/84#discussion_r751519248, but it looks like it's hard to find the original code, due to a rebase.

nickmoreton replied:
I added a second key to `settings.WORDPRESS_IMPORT_ITEMS_TO_CACHE` and a test for both saved in the TestImportHooksItemsCacheMethods() test. Let me know please if I have misunderstood you :)

nickmoreton replied:
@nimasmi I think this is the Issue and the following one that we talked about earlier today?

nimasmi replied:
See #96 for a solution to this.

nimasmi commented on wagtail_wordpress_import/test/tests/test_wordpress_item.py:
```diff
@@ -274,3 +274,68 @@ def test_items_dict_4(self):
         )
+
+
+class WordpressImporterTestsParsePostMeta(TestCase):
```
**Question**: What value of `settings.WORDPRESS_IMPORT_ITEMS_TO_CACHE` is used in these tests? Can we include a test that if there are two entries there, both are used? This is related to my comments on the ItemCache `.process()` and `._get_hook_handler_data()` methods.

nickmoreton replied:
Here there is no `settings.WORDPRESS_IMPORT_ITEMS_TO_CACHE` in play. It's testing the new method for populating `wp_post_meta` JSONField content. wp_post_meta is always saved do you think we should only save that if there is `settings.WORDPRESS_IMPORT_ITEMS_TO_CACHE`?

nickmoreton replied:
@nimasmi I think this is the Issue and the previous one that we talked about earlier today.

The class  ItemsCache approach is quite different in the Import Authors PR as I used your ideas for new the class TagsCache() for authors.

**Question** Should I pick out those changes and implement them here? There are quite a lot of changes that would complicate this PR though.

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
Update docs/import_hooks.md

Co-authored-by: Nick Smith <nick.smith@torchbox.com>

nickmoreton committed:
add a second item tag to test for

nimasmi committed:
Lint

nimasmi committed:
Add tests that multiple import hooks can be registered

nimasmi committed:
Permit multiple import hooks to be registered

nimasmi committed:
Remove deliberately failing test cases

nimasmi committed:
Fix test of import process, to ensure all handlers are called

nimasmi committed:
Merge pull request #96 from torchbox/import-hooks-multiple-handlers

Import hooks multiple handlers

nimasmi commented on wagtail_wordpress_import/importers/import_hooks.py:
```diff
@@ -0,0 +1,45 @@
+        registry = self._get_hook_handler_data()
+        for page in imported_pages:
+            for hook, (func, data_tag) in registry.items():
+                func(page, data_tag, self.__dict__)
```
**Suggestion**: Now we've merged the multiple-handlers changes, I would suggest that the data to be passed to the registered function should only take the data that it asks for. This would mean changing the example functions, too.

```suggestion
                func(page, data_tag, getattr(self, hook))
```

nickmoreton replied:
The example functions in the docs are working locally for me with no changes.
